tools/exploredb/main.go

```diff
@@ -95,143 +175,166 @@ func showBucketStats(dbNameWithPath string) {
-					maxKeySize = keyize
-				}
-				totalKeySize += uint64(len(k))
 				return nil
-			}); forEachErr != nil {
-				log.Errorf("could not process row %d for bucket: %s, %v", count, bName, forEachErr)
-				return forEachErr
+			}); viewErr2 != nil {
+				log.Errorf("could not get stats for bucket: %s, %v", bukName, viewErr2)
 			}
-			return nil
-		}); viewErr2 != nil {
-			log.Errorf("could not get stats for bucket: %s, %v", bName, viewErr2)
-			continue
-		}
-
-		if count != 0 {
-			averageValueSize := totalValueSize / count
-			averageKeySize := totalKeySize / count
-			fmt.Println("------ ", bName, " --------")
-			fmt.Println("NumberOfRows     = ", count)
-			fmt.Println("TotalBucketSize  = ", humanize.Bytes(totalValueSize+totalKeySize))
-			fmt.Println("KeySize          = ", humanize.Bytes(totalKeySize), "(min = "+humanize.Bytes(minKeySize)+", avg = "+humanize.Bytes(averageKeySize)+", max = "+humanize.Bytes(maxKeySize)+")")
-			fmt.Println("ValueSize        = ", humanize.Bytes(totalValueSize), "(min = "+humanize.Bytes(minValueSize)+", avg = "+humanize.Bytes(averageValueSize)+", max = "+humanize.Bytes(maxValueSize)+")")
-		}
+			stat := &bucketStat{
+				bucketName:     bukName,
+				noOfRows:       count,
+				totalKeySize:   totalKeySize,
+				totalValueSize: totalValueSize,
+				minKeySize:     minKeySize,
+				maxKeySize:     maxKeySize,
+				minValueSize:   minValueSize,
+				maxValueSize:   maxValueSize,
+			}
+			statsC <- stat
+		}(bName)
 	}
+	wg.Wait()
+	close(statsC)
 }
 
-func printBucketContents(dbNameWithPath string, rowLimit uint64, bucketName string) {
-	// get the keys within the supplied limit for the given bucket.
-	bucketNameInBytes := []byte(bucketName)
-	keys, sizes := keysOfBucket(dbNameWithPath, bucketNameInBytes, rowLimit)
-
-	// create a new KV Store.
-	dbDirectory := filepath.Dir(dbNameWithPath)
-	db, openErr := kv.NewKVStore(context.Background(), dbDirectory, &kv.Config{})
-	if openErr != nil {
-		log.Fatalf("could not open db, %v", openErr)
+func readStates(ctx context.Context, db *kv.Store, stateC chan<- *modifiedState, keys [][]byte, sizes []uint64) {
+	for rowCount, key := range keys {
+		st, stateErr := db.State(ctx, bytesutil.ToBytes32(key))
+		if stateErr != nil {
+			log.Errorf("could not get state for key : , %v", stateErr)
+		}
+		mst := &modifiedState{
+			state:     st,
+			key:       key,
+			valueSize: sizes[rowCount],
+			rowCount:  uint64(rowCount),
+		}
+		stateC <- mst
 	}
+	close(stateC)
+}
 
-	// dont forget to close it when ejecting out of this function.
-	defer func() {
-		closeErr := db.Close()
-		if closeErr != nil {
-			log.Fatalf("could not close db, %v", closeErr)
+func readStateSummary(ctx context.Context, db *kv.Store, stateSummaryC chan<- *modifiedStateSummary, keys [][]byte, sizes []uint64) {
+	for rowCount, key := range keys {
+		ss, ssErr := db.StateSummary(ctx, bytesutil.ToBytes32(key))
+		if ssErr != nil {
+			log.Errorf("could not get state summary for key : , %v", ssErr)
```

Comment:
A common pattern we use is the `WithError(error)` method with logrus. It adds the error as field metadata and makes it easier to query these logs in a deployed environment. 
```suggestion
			log.WithError(ssErr).Error("could not get state summary for key")
```

Reply:
cool. Will use this in mind for future too.